clippy_lints/src/utils/higher.rs

```diff
@@ -199,6 +199,18 @@ pub fn for_loop(expr: &hir::Expr) -> Option<(&hir::Pat, &hir::Expr, &hir::Expr)>
     None
 }
 
+/// Recover the essential nodes of a desugared if block
+/// `if cond { then } else { els }` becomes `(cond, then, Some(els))`
+pub fn if_block(expr: &hir::Expr) -> Option<(&hir::Expr, &hir::Expr, Option<&hir::Expr>)> {
+    if let hir::ExprKind::Match(ref cond, ref arms, hir::MatchSource::IfDesugar {..}) = expr.node {
+        let then = &arms[0].body;
+        let els = arms.get(1).map(|x| &*x.body);
+        Some((cond, then, els))
```

Comment:
You want to extract out the `DropTemps` from `cond` here; https://github.com/rust-lang/rust/pull/59288/files?file-filters%5B%5D=.rs#diff-64b696b0ef6ad44140e973801ed82b25R4150

Reply:
ah, dammit. I'll fix that once I've got the rest compiling